[시온] 미션 5: 보드에 위치 부여 및 점수 계산, 제출합니다.

ehdrhelr opened the pull request:
안녕하십니까.

코드스쿼드 2021 마스터즈 백엔드 과정 Shion 입니다.

이번 미션에서 세로로 중복된 Pawn들의 점수를 계산하기 위해 Board에서 이전 Rank를 저장하여 현재 Rank와 비교하는 방법으로 요구사항을 구현하려고 했는데 코드가 복잡해지고, 예상치 못한 오류가 지속적으로 발생하여 몇일동안 해결해보려고 고민하다가 실패하고 결국 자바지기님이 구현하신 코드를 참고해서 구현했습니다 ㅠㅠ

특이사항은 아래와 같고 코드 리뷰 부탁드립니다~

- Piece클래스가 Position정보를 갖으며, 자신의 점수를 계산하는 책임을 갖습니다.
- Position클래스는 자신과 위, 아래로 이웃한 Position정보를 알 수 있습니다.
- Board는 남아 있는 기물정보를 알 수 있고, 남아있는 각 기물들에게 점수를 요청하여 최종스코어를 구할 수 있습니다.

감사합니다!

Commit:
refactor : add Color and Type enum in Piece class

Replace literals related to color and representation with a use of enums.

Commit:
refactor : add createWhite(type) and createBlack(type)
to remove duplications

Commit:
refactor : add Rank class

1. Rank클래스를 이용하여 기존 4개의 리스트를 하나로 통합했습니다.

2. Board위의 모든 출력은 Piece 객체로 관리됩니다.

Commit:
test : add driving test codes checking count of pieces with specific color and type

Commit:
feat : add counting pieces with specific color and type

인자로 받은 color와 type을 동시에 만족하는 piece의 개수를 구할 수 있습니다.

Commit:
test : add driving test codes finding piece

입력된 위치에 존재하는 piece를 반환하는 기능을 추가하기 위한 테스트 코드를 먼저 작성한다.

Commit:
feat : add finding piece by input

1. 테스트 코드에서 입력으로 찾은 piece의 값들과 팩토리메서드로 만든 piece를 비교하기 위해 Piece클래스 안에 equals와 hashcode를 오버라이딩하였습니다.

2. piece를 찾기 위한 입력을 받기 위해 ChessGame클래스에 메소드를 추가하였습니다. (예외처리는 되어 있지 않습니다.)

Commit:
test : add driving test codes putting piece in position

Commit:
feat : add putting piece in position

1. 체스판을 빈칸으로 초기화하는 메소드 추가.

2. Rank의 결과를 가져오는 메소드를 하나로 통일.

Commit:
test : add driving test codes calculating score

Commit:
feat : add calculating score

남아있는 기물에 따라 점수를 계산할 수 있다.

Rank클래스는 한 Rank의 정보만 갖고 있기 때문에 Board클래스에서 상위 Rank의 정보를 저장하고 있다가 현재 Rank와 비교해서 중복된 Pawn을 찾습니다.

Commit:
feat : add sorting remained pieces in order

색깔 별로 남아있는 기물들을 점수가 높은 순은 순서에서 낮은 순으로 정렬한다.

Commit:
style : use static import to access constants

코드를 간결하게 하기 위해 클래스를 import하여 사용합니다.

Commit:
refactor : change for to lambda

코드를 간결하게 만들기 위해 Piece클래스의 representation을 가져오는 메서드들을 하나로 통일하였습니다.

Board클래스의 getRankResult를 통일한 메소드를 이용해 간소화했습니다.

Commit:
refactor : add Position class

중복을 제거하기 위해 입력된 위치의 file과 rank를 Position객체에 담아서 관리합니다.

Commit:
refactor : use lambda and import to make codes simplier

Commit:
fix : add calculating score for duplication of pawns on file

1. 잘못된 변수명을 바로 고쳤습니다.

2. 이전의 rank와 현재 rank를 비교하여 세로로 중복되는 pawn의 점수를 반영합니다.

Commit:
fix : remove wrong calculation

0. 기존 점수계산 방식은 PAWN의 위치에 따라 예상치 못한 값을 도출하는 심각한 오류를 갖고 있어 자바지기님의 코드를 참조했습니다.

1. Piece가 Position을 갖기 시작하고, Position을 통해 자신의 점수를 계산합니다.

2. Position은 자신과 같은 file의 위, 아래의 위치를 반환하는 메소드를 갖고 있고, 이를 통해서 세로로 인접한 PAWN의 점수를 계산합니다.

Commit:
test : refactor test code comparing lists

SANGCO commented on src/test/java/chess/BoardTest.java:
```diff
@@ -1,7 +1,13 @@
 package chess;
```
패키지 정리가 필요해 보입니다!
net.honux.chess는 없애고 net.sion.chess 뭐 이런식으로 본인의 패키지를 만들고
net.sion.chess 하위에 모든 클래스들을 위치 시켜 봅시다.

ehdrhelr replied:
리뷰해주신대로 패키지구조를 정리했습니다~
훨씬 깔끔하고 안정적이네요!

SANGCO commented on src/main/java/chess/ChessGame.java:
```diff
@@ -1,5 +1,7 @@
+
 import java.util.Scanner;
 
 public class ChessGame {
```
try-with-resources 관련되어서 학습하고 적용해보면 좋을거 같아요~
Scanner와 같이 AutoCloseable을 구현한 클래스의 경우 
[try-with-resources](https://www.baeldung.com/java-try-with-resources)를 사용해서 쉽게 리소스를 해제할 수 있답니다.

ehdrhelr replied:
링크 감사합니다~
리뷰해주신대로 try-with-resources 적용해서 변경했습니다!
리소스를 사용하고 반환하는 습관을 들여야겠습니다 ㅎㅎ

SANGCO commented on src/main/java/chess/pieces/Piece.java:
```diff
@@ -1,99 +1,128 @@
+    @Override
+    public String toString() {
+        return type.name();
     }
```
요 세개가 제일 밑으로 와야 합니다.

ehdrhelr replied:
enum이 그래도 클래스니까 메소드보다 더 무겁다고(?) 생각하여 나름 최대한 밑으로 보낸 것이었는데 맨 밑으로 가야하는군요!
수정하였습니다~

SANGCO commented on src/main/java/chess/Board.java:
```diff
@@ -5,103 +5,176 @@
+        for (int i = 0; i < MAX_RANK; i++) {
+            board.add(new Rank());
+        }
+    }
```
왜 디폴트 접근제한자를 선택 하셨나요~

SANGCO replied:
Board 클래스에 board 필드 뭔가 어색하네요.
ranks 아니면 rankList 어떤가요?

ehdrhelr replied:
접근제한자는 신경 못쓰고 지나쳤던 것 같습니다 ㅜㅜ 캡슐화와 객체 간의 협력을 마음에 새기겠습니다...
board 필드가 랭크들이나 랭크리스트지만 결국 체스판을 나타낸다고 생각해서 좀 더 추상적인 표현으로 작명했습니다만, 다시 생각해보니 ranks나 rankList처럼 좀 더 직관적으로 알 수 있게 구체적으로 표현하는게 더 좋은 것 같습니다!
리뷰 참고하여 수정했습니다~!

SANGCO commented on src/main/java/chess/Board.java:
```diff
@@ -5,103 +5,176 @@
-    /* 체스판을 초기화한다. */
     public void initialize() {
-        clearAll();
-
         addBlackPieces();
         addBlackPawns();
+        addBlank();
         addWhitePawns();
         addWhitePieces();
     }
```
초기화 메서드를 위쪽으로 이동을 시키던지 생성자에서 이 작업을 하는건 어떨까요?

ehdrhelr replied:
객체 생성 시 자동으로 초기화해주는 것이 흐름상 더 좋을 것 같습니다 ㅎㅎ
리뷰를 참고하여 initialize를 생성자에 넣고 관련 메소드들을 바로 밑으로 이동시켰습니다!

SANGCO commented on src/main/java/chess/Board.java:
```diff
@@ -5,103 +5,176 @@
+        return blankRank;
+    }
+
     private void addBlackPieces() {
```
`addBlackPieces()`, `addBlackPieces()`, `addBlank ()`, `addWhitePawns ()`, `addWhitePieces ()`
이 메서드가 하는 일을 `Rank` 클래스가 할 수는 없을까요?
정적 팩토리 메서드를 찾아서 학습하고 적용해 봅시다.

ehdrhelr replied:
Board에서 모든 기물을 관리해야된다고 생각해서 초기화까지 Board에서 해주었는데, 각 Rank의 초기화는 Rank에게 맡기고 Board에서 요청하는게 더 맞는 이야기 같습니다!
코드도 더 깔끔해진 것 같네요 ㅎㅎㅎ

SANGCO commented on src/main/java/chess/Board.java:
```diff
@@ -5,103 +5,176 @@
-    public static final int MAX_PAWN = 8;
+    public static final int MAX_RANK = 8;
+    public static final int MAX_FILE = 8;
+    public static final int RANK_1 = 0;
+    public static final int RANK_2 = 1;
+    public static final int RANK_3 = 2;
+    public static final int RANK_4 = 3;
+    public static final int RANK_5 = 4;
+    public static final int RANK_6 = 5;
+    public static final int RANK_7 = 6;
+    public static final int RANK_8 = 7;
+    public static final int FILE_A = 0;
+    public static final int FILE_B = 1;
+    public static final int FILE_C = 2;
+    public static final int FILE_D = 3;
+    public static final int FILE_E = 4;
+    public static final int FILE_F = 5;
+    public static final int FILE_G = 6;
+    public static final int FILE_H = 7;
```
밑에 리뷰 드린게 반영이 되면 이 부분도 좀 정리가 될거 같군요.

ehdrhelr replied:
말씀하신대로 잘 정리가 되었습니다 ㅎㅎㅎ

SANGCO reviewed (changes requested):
시온님 구현 잘하셨네요. 고생하셨습니다!
리뷰 남겼으니 확인해 보시고 질문 있으시면 댓글 부탁드립니다~

Commit:
style : change package structure

1. 사용하지 않는 패키지(net.honux.chess)를 제거하고 Shion패키지 생성하여 클래스를 관리합니다.

Commit:
refactor : use try-with-resources with Scanner to close resource

Commit:
style : send overriding methods to bottom

Commit:
refactor : change access identifier and variable name

1. 멤버변수를 private으로, 생성자를 public으로 변경하였습니다.

2. 변수 이름을 조금 더 구체적인 이름으로 변경하였습니다.

Commit:
style : send methods related to initializing board upper

1. Board 생성자에 초기화 메서드를 추가하여, 객체 생성 시 초기화도 함께 진행합니다.

2. 가독성을 위해 초기화 관련 메서드를 생성자 아래로 이동시켰습니다.

Commit:
refactor : add static factory methods in Rank

이제 Rank는 정적 팩토리 메소드를 통해 1부터 8까지 각 Rank의 초기 기물 배치 역할을 수행합니다.

Board는 Rank에게 메세지를 보내 각 Rank의 초기 정보를 받아서 알맞게 배치합니다.

ghost commented:
친절하게 해주신 리뷰 덕분에 코드를 개선하였으니 참고해주시면 감사하겠습니다~!

SANGCO reviewed (approved):
수정 잘하셨네요. 
고생하셨어요~
